### PR TITLE
Filter out intrinsics if we have other import candidates to suggest

### LINKED ISSUE #97618: E0425: Don't recommend the use of intrinsics first
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dc1180286b0a5d9e0f1dbac02ccdb7b3

```rust
fn main() {
    let sz = size_of::<usize>();
    println!("Hello, world! A usize is {}", sz);
}
```

The current output is:

```
error[[E0425]](https://doc.rust-lang.org/stable/error-index.html#E0425): cannot find function `size_of` in this scope
 --> src/main.rs:2:14
  |
2 |     let sz = size_of::<usize>();
  |              ^^^^^^^ not found in this scope
  |
help: consider importing one of these items
  |
1 | [use core::intrinsics::size_of;](https://play.rust-lang.org/#)
  |
1 | [use core::mem::size_of;](https://play.rust-lang.org/#)
  |
1 | [use memoffset::__priv::mem::size_of;](https://play.rust-lang.org/#)
  |
1 | [use pin_utils::core_reexport::intrinsics::size_of;](https://play.rust-lang.org/#)
  |
    and 3 other candidates
```

Ideally the output should look like:

```
error[[E0425]](https://doc.rust-lang.org/stable/error-index.html#E0425): cannot find function `size_of` in this scope
 --> src/main.rs:2:14
  |
2 |     let sz = size_of::<usize>();
  |              ^^^^^^^ not found in this scope
  |
help: consider importing one of these items
  |
1 | [use core::mem::size_of;](https://play.rust-lang.org/#)
  |
1 | [use core::intrinsics::size_of;](https://play.rust-lang.org/#)
 |
1 | [use memoffset::__priv::mem::size_of;](https://play.rust-lang.org/#)
  |
1 | [use pin_utils::core_reexport::intrinsics::size_of;](https://play.rust-lang.org/#)
  |
    and 3 other candidates
```

In cases where intrinsics have been stabilised in `core`|`std` it would be useful to suggest these to users first.
Especially as using intrinsics generates another `rustc` error that tells you to prefer stabilized functions in `std`.
This would hopefully make handling this error easier for new rust users coming from languages like C/C++ and looking for the equivalent of `sizeof()`

The output is the same on stable, beta and nightly

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes #97618

Also open to just sorting these candidates to be last. Pretty easy to modify the code to do that, too.

### CODE REVIEW

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 1]
```diff
@@ -396,6 +396,18 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
             })
             .collect::<Vec<_>>();
         let crate_def_id = CRATE_DEF_ID.to_def_id();
+        // Try to filter out intrinsics candidates, as long as we have
+        // some other candidates to suggest.
+        let intrinsic_candidates: Vec<_> = candidates
+            .drain_filter(|sugg| {
+                let path = path_names_to_string(&sugg.path);
+                path.starts_with("core::intrinsics::") || path.starts_with("std::intrinsics::")
```
Comment (oli-obk) [inline compiler/rustc_resolve/src/late/diagnostics.rs:404]: You could add a `#[diagnostic_item]` attribute to the intrinsics `extern` block and check that the parent of the suggestion is not that specific diagnostic item.

**Comment by oli-obk** [comment]
just plain filtering them out sgtm

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_resolve/src/late/diagnostics.rs:404]: do I have access to diagnostic items in rustc_resolve? i don't have a tcx here, i don't think.

Comment (oli-obk) [reply compiler/rustc_resolve/src/late/diagnostics.rs:404]: oh right :D

**Pull request merged**